### ISSUE TITLE
Support HTTP/2 on the client side

### ISSUE THREAD

**Issue by Geal** [apollographql/router#1364]
this might be already working, but we have to make sure it"s there

**Comment by garypen** [comment]
I just checked with curl and it's not enabled. It's an axum feature (http2) and not enabled by default.

**Comment by garypen** [comment]
Enabling the feature in axum doesn't help because we are building our connections from the low level hyper primitives:
```
let connection = Http::new()
    .http1_keep_alive(true)
    .serve_connection(stream, app);
```
I'll look into fixing this to add http2 support.

**Comment by garypen** [comment]
Ok. We can ignore that previous comment. It turns out that curl `-http2` flag works by requesting an upgrade from HTTP1.1 and this is [not supported by hyper](https://github.com/hyperium/hyper/issues/2411)

Luckily curl has the `-http2-prior-knowledge` flag which doesn't attempt to upgrade from HTTP1.1. If I run my curl command using that flag against an unmodified router, i.e.: current mainline, then this works fine.
```
curl -v -k --http2-prior-knowledge \
    --header 'content-type: application/json' \
    --url 'http://localhost:4000/' \
    --data '{"query":"query ExampleQuery($first: Int) {\n  topProducts(first: $first) {\n    name\n  }\n}","variables":{"first":1}}'
*   Trying 127.0.0.1:4000...
* Connected to localhost (127.0.0.1) port 4000 (#0)
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x141811800)
> POST / HTTP/2
> Host: localhost:4000
> user-agent: curl/7.79.1
> accept: */*
> content-type: application/json
> content-length: 119
> 
* We are completely uploaded and fine
* Connection state changed (MAX_CONCURRENT_STREAMS == 4294967295)!
< HTTP/2 200 
< content-type: application/json
< content-length: 43
< vary: origin
< vary: access-control-request-method
< vary: access-control-request-headers
< date: Wed, 06 Jul 2022 15:55:23 GMT
< 
* Connection #0 to host localhost left intact
{"data":{"topProducts":[{"name":"Table"}]}}
```
I'm going to close this issue since we do support HTTP2 already.

**Comment by Geal** [comment]
Did you check that it works with a browser?

**Comment by garypen** [comment]
I did. It works when the connection is over https. I note that when talking over http, it uses 1.1.

I think the reason for this is the issue that I referenced in hyper where HTTP1.1 upgrade isn't implemented. I don't think that will be a problem in "real", i.e.: production scenarios, but it's something to bear in mind if you get "surprising" results when you are checking if things are working or not.
